samza-azure/src/main/java/org/apache/samza/system/eventhub/admin/EventHubSystemAdmin.java

```diff
@@ -0,0 +1,199 @@
+        });
+
+    } finally {
+
+      // Closing clients
+      eventHubClients.forEach((streamName, client) -> client.close(DEFAULT_SHUTDOWN_TIMEOUT_MILLIS));
+      eventHubClients.clear();
+    }
+
+    return requestedMetadata;
+  }
+
+  private EventHubClientManager getOrCreateStreamEventHubClient(String streamName) {
+    if (!eventHubClients.containsKey(streamName)) {
+      LOG.debug(String.format("Creating EventHubClient for Stream=%s", streamName));
+
+      EventHubClientManager eventHubClientManager = eventHubClientManagerFactory
+              .getEventHubClientManager(systemName, streamName, eventHubConfig);
+
+      eventHubClientManager.init();
+      eventHubClients.put(streamName, eventHubClientManager);
+    }
+    return eventHubClients.get(streamName);
+  }
+
+  private Map<Partition, SystemStreamPartitionMetadata> getPartitionMetadata(String streamName, String[] partitionIds) {
+    EventHubClientManager eventHubClientManager = getOrCreateStreamEventHubClient(streamName);
+    Map<Partition, SystemStreamPartitionMetadata> sspMetadataMap = new HashMap<>();
+    Map<String, CompletableFuture<EventHubPartitionRuntimeInformation>> ehRuntimeInfos = new HashMap<>();
+
+    for (String partition : partitionIds) {
+      CompletableFuture<EventHubPartitionRuntimeInformation> partitionRuntimeInfo = eventHubClientManager
+              .getEventHubClient()
+              .getPartitionRuntimeInformation(partition);
+
+      ehRuntimeInfos.put(partition, partitionRuntimeInfo);
+    }
+
+    ehRuntimeInfos.forEach((partitionId, ehPartitionRuntimeInfo) -> {
+        try {
+          long timeoutMs = eventHubConfig.getRuntimeInfoWaitTimeMS(systemName);
+          EventHubPartitionRuntimeInformation ehPartitionInfo = ehPartitionRuntimeInfo.get(timeoutMs, TimeUnit.MILLISECONDS);
+
+          String startingOffset = EventHubSystemConsumer.START_OF_STREAM;
+          String newestOffset = ehPartitionInfo.getLastEnqueuedOffset();
+          String upcomingOffset = getNextOffset(newestOffset);
+          SystemStreamPartitionMetadata sspMetadata = new SystemStreamPartitionMetadata(startingOffset, newestOffset,
+                  upcomingOffset);
+
+          Partition partition = new Partition(Integer.parseInt(partitionId));
+
+          sspMetadataMap.put(partition, sspMetadata);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+          String msg = String.format(
+                  "Error while fetching EventHubPartitionRuntimeInfo for System:%s, Stream:%s, Partition:%s",
+                  systemName, streamName, partitionId);
+          throw new SamzaException(msg);
+        }
+      });
+    return sspMetadataMap;
+  }
+
+  public static Integer compareOffsets(String offset1, String offset2) {
+    if (offset1 == null || offset2 == null) {
+      return null;
+    }
+    try {
+      if (offset1.equals(EventHubSystemConsumer.END_OF_STREAM)) {
```

Comment:
we appear to be not comparing the actual offsets (we appear to be handling cases only where offset1 and offset2 are either null or EOS. Is that intentional? Should we convert them to Long and compare?